### ISSUE TITLE
Implement the gradient for depthwiseConv2d.

### ISSUE THREAD

**Issue by dsmilkov** [tensorflow/tfjs#71]
_From @nsthorat on March 24, 2018 18:0_

Currently we don't have the gradient written. It should be strictly easier than conv2d, so should be easy.

This is important for training in browsers where compute is limited.

_Copied from original issue: tensorflow/tfjs-core#892_

**Comment by jaxball** [comment]
Is someone working on this? I'd like to get the gradients for MobileNet to perform finetuning in the browser and visualize features on the original image like Deepdream does. 

After some digging through, I confirm the need for this ticket as part of my research and when I printed out `filteredTape` in `Engine.prototype.gradients()`, I also see some _undefined_ layers interlaced between consecutive pairs of _reshape_ ops (see pic). @nsthorat I see in https://github.com/tensorflow/tfjs/issues/108#issue-311720751 that the team is working on this, so I was wondering if I could be pointed somewhere. 
![screen shot 2018-04-19 at 9 43 42 pm](https://user-images.githubusercontent.com/8799208/39027029-0925cf60-441e-11e8-8da3-ac1220a09b63.png)


Also, this comment is a followup to https://github.com/tensorflow/tfjs/issues/153. Nikhil you said 

> We don't have gradients defined for depthwise convs, so MobileNet gradients should have worked in the old way. 

I'm not sure what you mean by the old way. If you're referring to getting the gradients, that's precisely what we want. How do you go about it?

**Comment by dsmilkov** [comment]
You need the gradients for depthwise conv in order to analyze MobileNet's gradients. We will add this early next week. Stay tuned. In the meantime if anyone else is interested in taking a stab at this, I'll be happy to review.

**Comment by jaxball** [comment]
Gotcha. If I wanted to take a quick jab at it, what are some files I should specifically look out for? 
My project deadline is next week so I'd like to have it added by early next week. I'd love to help given some pointers! 

Thanks,
Jason

**Comment by dsmilkov** [comment]
Let's take conv2d since it's very similar to depthwise conv2d. [Here are the lines](https://github.com/tensorflow/tfjs-core/blob/master/src/ops/conv.ts#L190-L209) that specify forward mode and gradients for conv2d. That op then uses 2 kernel/backend ops [specified here](https://github.com/tensorflow/tfjs-core/blob/master/src/kernels/backend_webgl.ts#L810-L819).

DepthwiseConv2d is [analogous](https://github.com/tensorflow/tfjs-core/blob/master/src/ops/conv.ts#L482-L491) but it's missing the gradients. You'd have to add `depthwiseConv2dDerInput` and `depthwiseConv2DDerFilter` to the backend. The math should be simpler than the `conv2dDerInput` and `conv2dDerFilter`. Hope this helps!

**Comment by carlthome** [comment]
Status on this? I'd really love to be able to calculate gradients for doing saliency maps.
